src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs

```diff
@@ -72,18 +72,14 @@ private static IArgumentConstraint CreateEqualityConstraint(object expressionVal
             return new EqualityArgumentConstraint(expressionValue);
         }
 
-        private static object InvokeExpression(Expression expression)
-        {
-            return Expression.Lambda(expression).Compile().DynamicInvoke();
-        }
-
         private IArgumentConstraint GetArgumentConstraintFromExpression(Expression expression)
         {
             object expressionValue = null;
 
             var trappedConstraints = this.argumentConstraintTrapper.TrapConstraints(() =>
             {
-                expressionValue = InvokeExpression(expression);
+                // in this ParsedArgumentExpression scenario, we get benefits from the same optimizations as ParsedCallExpression
+                expressionValue = ParsedCallExpression.FastEvaluate(expression);

```

Comment:
Another way to look at it: supposedly this optimizing away some dynamic invoke() thing, but what does that mean! It just means we are optimizing away bits of the compiler/scripting for evaluating, not optimizing away the script that is to be evaluated - it's still all evaluated.